services-camel/src/main/java/edu/unc/lib/dl/services/camel/longleaf/RegisterToLongleafProcessor.java

```diff
@@ -0,0 +1,251 @@
+        );
+
+    private String longleafBaseCommand;
+
+    private RepositoryObjectLoader repoObjLoader;
+
+    private FcrepoClient fcrepoClient;
+
+    /**
+     * The exchange here is expected to be a batch message with one or more
+     */
+    @SuppressWarnings("unchecked")
+    @Override
+    public void process(Exchange exchange) throws Exception {
+        Message aggrMsg = exchange.getIn();
+
+        // Key is the digest key, value is a map of storage uri to digest value
+        Map<String, Map<URI, String>> digestsMap = new HashMap<>();
+        Map<URI, String> md5Map = new HashMap<>();
+        Map<URI, String> sha1Map = new HashMap<>();
+        digestsMap.put("md5", md5Map);
+        digestsMap.put("sha1", sha1Map);
+
+        List<String> messages = aggrMsg.getBody(List.class);
+        for (String fcrepoUri : messages) {
+            try {
+                PID pid = PIDs.get(fcrepoUri);
+                BinaryObject binObj = repoObjLoader.getBinaryObject(pid);
+
+                String md5 = trimFedoraDigest(binObj.getMd5Checksum(), ":");
+                String sha1 = trimFedoraDigest(binObj.getSha1Checksum(), ":");
+                URI storageUri = binObj.getContentUri();
+
+                if (md5 != null) {
+                    md5Map.put(storageUri, md5);
+                }
+                if (sha1 != null) {
+                    sha1Map.put(storageUri, sha1);
+                }
+                if (md5 == null && sha1 == null) {
+                    sha1Map.put(storageUri, calculateSha1(pid));
+                }
+            } catch (Exception e) {
+                log.error("Failed to add {} to batch for regisration to longleaf", fcrepoUri, e);
+            }
+        }
+
+        registerFiles(digestsMap);
+    }
+
+    private String calculateSha1(PID pid) {
+        try (FcrepoResponse response = fcrepoClient.head(pid.getRepositoryUri())
+                .addHeader(FedoraHeaderConstants.WANT_DIGEST, "sha")
+                .perform()) {
+            String digestVal = response.getHeaderValue(FedoraHeaderConstants.DIGEST);
+            if (digestVal == null) {
+                throw new ServiceException("Failed to calculate sha1 for " + pid.getRepositoryPath());
+            } else {
+                return trimFedoraDigest(digestVal, "=");
+            }
+        } catch (IOException e) {
+            throw new ServiceException(e);
+        } catch (FcrepoOperationFailedException e) {
+            ClientFaultResolver.resolve(e);
+            return null;
+        }
+    }
+
+    /**
+     * Executes longleaf register command for a batch of files with digests
+     *
+     * @param digestsMap mapping of digest algorithms to paths plus digest values
+     */
+    private void registerFiles(Map<String, Map<URI, String>> digestsMap) {
+        long start = System.currentTimeMillis();
+
+        StringBuilder sb = new StringBuilder();
+        final int[] cntArray = { 0 };
```

Comment:
Why isn't this just a regular int?

Reply:
because its being incremented inside of nested `forEach` blocks. Variables used inside of these blocks must be final, but for primatives like ints you can't modify them if they are final. So instead its incrementing a value inside of a final array. Other methods would be to use an AtomicInteger or a MutableInt, perhaps one of those would be less confusing. I don't know if there is an performance difference.